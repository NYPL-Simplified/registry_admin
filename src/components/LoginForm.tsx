import React, { FormEvent, useContext, useState } from 'react';
import Cookies from 'js-cookie';
import {
  Box,
  Button,
  Form,
  FormField,
  Heading,
  HelperErrorText,
  TextInput,
} from '@nypl/design-system-react-components';

import { LOGIN } from '../constants';
import useTokenContext from '../context/tokenContext';

interface LoginFormProps {
  error: string;
  setError: (error: string) => void;
}

const backgroundStyles = {
  background: 'section.blogs.primary',
  margin: 'auto',
  marginTop: 'xxl',
  paddingBottom: 'xl',
  paddingTop: 'l',
  width: '400px',
};
const formFieldStyles = { margin: 'auto', width: '75%' };
const errorStyles = {
  color: 'ui.error.primary',
  paddingTop: 'xs',
  textAlign: 'center',
};

const LoginForm = ({ error, setError }: LoginFormProps) => {
  const [username, setUsername] = useState<string>('');
  const [password, setPassword] = useState<string>('');

  const { setAccessToken } = useTokenContext();

  const login = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    // Create a FormData object to store the username and password for sending
    // with the login request.
    const body = new FormData();
    body.append('username', username);
    body.append('password', password);

    fetch(LOGIN as RequestInfo | URL, {
      method: 'POST',
      body,
    })
      .then((response) => {
        if (response.ok) {
          // Reset setError to an empty string just in case it is set to an old
          // error.
          setError('');
          return response.json();
        }
        if (response.status >= 400 && response.status < 502) {
          // This error will be displayed to the user.
          setError('Your username or password is incorrect. Please try again.');
        }
        throw new Error(
          'There was a problem authenticating this user in LoginForm.tsx.'
        );
      })
      .then((data) => {
        // After a successful request, store the two tokens received in their
        // their appropriate places: the access token in context and the refresh
        // token in cookies.
        setAccessToken(data.access_token);
        Cookies.set('refreshToken', data.refresh_token, {
          sameSite: 'lax',
          secure: true,
        });
      })
      .catch((err) => {
        console.log(err.message);
      });
  };

  return (
    <Box __css={backgroundStyles}>
      <Heading level='two' size='tertiary' __css={{ textAlign: 'center' }}>
        Log In
      </Heading>
      <Form id='login' onSubmit={(e: FormEvent<HTMLFormElement>) => login(e)}>
        <FormField>
          <TextInput
            id='username'
            isRequired
            labelText='Username'
            onChange={(e) => setUsername(e.target.value)}
            showRequiredLabel={false}
            type='text'
            __css={formFieldStyles}
          />
        </FormField>
        <FormField>
          <TextInput
            id='password'
            isRequired
            labelText='Password'
            onChange={(e) => setPassword(e.target.value)}
            showRequiredLabel={false}
            type='password'
            __css={formFieldStyles}
          />
        </FormField>
        <FormField>
          <Button id='submit' type='submit' __css={formFieldStyles}>
            Submit
          </Button>
        </FormField>
      </Form>
      <HelperErrorText
        ariaAtomic={false}
        ariaLive='assertive'
        id='loginError'
        text={error && error}
        __css={errorStyles}
      />
    </Box>
  );
};

export default LoginForm;
