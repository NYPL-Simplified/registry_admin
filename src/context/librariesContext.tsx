import React, {
  createContext,
  Dispatch,
  useContext,
  useEffect,
  useState,
} from 'react';
export interface LibraryData {
  areas: {
    focus: string[];
    service: string[];
  };
  basic_info: {
    description: string;
    internal_urn: string;
    name: string;
    number_of_patrons: string;
    online_registration: string;
    pls_id: string;
    short_name: string;
    timestamp: string;
  };
  stages: {
    library_stage: 'testing' | 'production' | 'cancelled';
    registry_stage: 'testing' | 'production' | 'cancelled';
  };
  urls_and_contact: {
    authentication_url: string;
    contact_email: string | null;
    contact_validated: string;
    copyright_email: string | null;
    copyright_validated: string;
    help_email: string | null;
    help_validated: string;
    opds_url: string;
    web_url: string;
  };
  uuid: string;
}

type LibrariesContextValues = {
  libraries: LibraryData[];
  setLibraries: Dispatch<React.SetStateAction<LibraryData[]>>;
  setUpdatedLibrary: Dispatch<React.SetStateAction<LibraryData | undefined>>;
};

type LibrariesProviderProps = {
  children: React.ReactNode;
};

export const LibrariesContext = createContext<
  LibrariesContextValues | undefined
>(undefined);

export const LibrariesProvider = ({ children }: LibrariesProviderProps) => {
  const [updatedLibrary, setUpdatedLibrary] = useState<
    LibraryData | undefined
  >();
  const [libraries, setLibraries] = useState<LibraryData[]>([]);

  useEffect(() => {
    if (updatedLibrary) {
      // Find the index of the library with the same uuid as the one
      // that was recently updated.
      const indexOfElementToReplace = libraries.findIndex(
        (library) => library.uuid === updatedLibrary.uuid
      );

      // Replace the library at that index with the updatedLibrary.
      const updatedLibraries = libraries.slice();
      updatedLibraries.splice(indexOfElementToReplace, 1, updatedLibrary);

      // Set libraries to the updated list.
      setLibraries(updatedLibraries);
    }
  }, [updatedLibrary]);

  return (
    <LibrariesContext.Provider
      value={{ libraries, setLibraries, setUpdatedLibrary }}
    >
      {children}
    </LibrariesContext.Provider>
  );
};

const useLibrariesContext = () => {
  const context = useContext(LibrariesContext);
  if (typeof context === 'undefined') {
    throw new Error('useLibraries must be used within a LibraryProvider');
  }
  return context;
};

export default useLibrariesContext;
