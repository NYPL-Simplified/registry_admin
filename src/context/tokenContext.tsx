import React, { createContext, Dispatch, useState } from 'react';

type TokenContextValues = {
  accessToken: string;
  setAccessToken: Dispatch<React.SetStateAction<string>>;
};

type TokenProviderProps = {
  children: React.ReactNode;
};

export const TokenContext = createContext<TokenContextValues | undefined>(
  undefined
);

export const TokenProvider = ({ children }: TokenProviderProps) => {
  const [accessToken, setAccessToken] = useState<string>('');

  return (
    <TokenContext.Provider value={{ accessToken, setAccessToken }}>
      {children}
    </TokenContext.Provider>
  );
};

const useTokenContext = () => {
  const context = React.useContext(TokenContext);
  if (typeof context === 'undefined') {
    throw new Error('useLibraries must be used within a LibraryProvider');
  }
  return context;
};

export default useTokenContext;
