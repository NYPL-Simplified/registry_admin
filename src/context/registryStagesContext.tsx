import React, { createContext, Dispatch, useState } from 'react';
import { LibraryStage } from '../components/LibraryDetails';

export type RegistryStagesContextValues = {
  registryStages: Record<string, LibraryStage>;
  setRegistryStages: Dispatch<
    React.SetStateAction<Record<string, LibraryStage>>
  >;
};

type RegistryStagesProviderProps = {
  children: React.ReactNode;
};

export const RegistryStagesContext =
  createContext<RegistryStagesContextValues | null>(null);

export const RegistryStagesProvider = ({
  children,
}: RegistryStagesProviderProps) => {
  const [registryStages, setRegistryStages] = useState<
    Record<string, LibraryStage>
  >({});

  console.log('registryStages -->', registryStages);

  return (
    <RegistryStagesContext.Provider
      value={{ registryStages, setRegistryStages }}
    >
      {children}
    </RegistryStagesContext.Provider>
  );
};
