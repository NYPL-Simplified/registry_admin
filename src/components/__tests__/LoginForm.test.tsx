import * as React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';

import LoginForm from '../LoginForm';
import { TokenContext } from '../../context/tokenContext';

const setAccessTokenMock = jest.fn();
const mockFormData = new FormData();
mockFormData.append('username', 'mockUser');
mockFormData.append('password', 'mockPassword');

const renderLoginFormWithContext = (
  accessToken: string,
  setAccessToken = setAccessTokenMock
) => {
  return render(
    <TokenContext.Provider value={{ accessToken, setAccessToken }}>
      <LoginForm />
    </TokenContext.Provider>
  );
};

const renderLoginForSnapshot = (
  accessToken: string,
  setAccessToken = setAccessTokenMock
) => {
  return (
    <TokenContext.Provider value={{ accessToken, setAccessToken }}>
      <LoginForm />
    </TokenContext.Provider>
  );
};

describe('LoginForm', () => {
  beforeEach(() => {
    renderLoginFormWithContext('');
  });

  it('renders form components', () => {
    const usernameInput = screen.getByRole('textbox');
    const passwordInput = screen.getByLabelText(/password/i);

    expect(usernameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();

    const submitButton = screen.getByRole('button');

    expect(submitButton).toHaveTextContent(/submit/i);
  });
});

describe('LoginForm, succcessful API call', () => {
  it('makes a call to the appropriate endpoint when a user submits', async () => {
    (global as any).fetch = jest.fn().mockReturnValueOnce(
      Promise.resolve({
        json: () =>
          Promise.resolve({
            access_token: 'mockAccessToken',
            refresh_token: 'mockRefreshToken',
          }),
        ok: true,
        status: 200,
      })
    );
    renderLoginFormWithContext('');
    const usernameInput = screen.getByRole('textbox');
    const passwordInput = screen.getByLabelText(/password/i);
    const submitButton = screen.getByRole('button');

    fireEvent.change(usernameInput, { target: { value: 'mockUser' } });
    fireEvent.change(passwordInput, { target: { value: 'mockPassword' } });
    fireEvent.click(submitButton);

    expect(fetch).toHaveBeenCalledWith(process.env.LOGIN, {
      method: 'POST',
      body: mockFormData,
    });

    await waitFor(() =>
      expect(
        screen.queryByText(/your username or password is incorrect/i)
      ).not.toBeInTheDocument()
    );
  });
});

describe('LoginForm, unsuccessful API call', () => {
  it('shows the user an error if the they cannot be authenticated', async () => {
    const log = jest.spyOn(console, 'log');

    (global as any).fetch = jest.fn().mockReturnValueOnce(
      Promise.resolve({
        status: 401,
      })
    );
    renderLoginFormWithContext('');
    const usernameInput = screen.getByRole('textbox');
    const passwordInput = screen.getByLabelText(/password/i);
    const submitButton = screen.getByRole('button');

    fireEvent.change(usernameInput, { target: { value: 'mockUser' } });
    fireEvent.change(passwordInput, { target: { value: 'mockPassword' } });
    fireEvent.click(submitButton);

    expect(fetch).toHaveBeenCalledWith(
      'https://qa-libraryregistry.librarysimplified.org/admin/log_in/jwt',
      { method: 'POST', body: mockFormData }
    );

    await waitFor(() => {
      expect(log).toHaveBeenCalledWith(
        'There was a problem authenticating this user in LoginForm.tsx.'
      );
      expect(
        screen.getByText(/your username or password is incorrect/i)
      ).toBeInTheDocument();
    });
  });
});

describe('Login Snapshot', () => {
  it('renders the UI snapshot correctly', () => {
    const loginPage = renderer.create(renderLoginForSnapshot('')).toJSON();

    expect(loginPage).toMatchSnapshot();
  });
});
