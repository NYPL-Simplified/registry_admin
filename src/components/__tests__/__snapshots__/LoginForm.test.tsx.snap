// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Snapshot renders the UI snapshot correctly 1`] = `
<div
  className="css-jgblkz"
>
  <h2
    className="chakra-heading css-1iatin6"
  >
    Log In
  </h2>
  <form
    aria-label="form"
    className="css-1xdhyk6"
    id="login"
    onSubmit={[Function]}
  >
    <div
      className="css-fadkce"
      id="login-parent"
    >
      <div
        className="css-l5w1vc"
        id="login-child0"
      >
        <div
          className="css-5eatfv"
          id="username-wrapper"
        >
          <label
            className="css-1xdhyk6"
            htmlFor="username"
            id="username-label"
          >
            Username
          </label>
          <div
            className="css-79elbk"
          >
            <input
              aria-required={true}
              autoComplete={null}
              className="chakra-input css-0"
              disabled={false}
              id="username"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              step={null}
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        className="css-l5w1vc"
        id="login-child1"
      >
        <div
          className="css-5eatfv"
          id="password-wrapper"
        >
          <label
            className="css-1xdhyk6"
            htmlFor="password"
            id="password-label"
          >
            Password
          </label>
          <div
            className="css-79elbk"
          >
            <input
              aria-required={true}
              autoComplete={null}
              className="chakra-input css-0"
              disabled={false}
              id="password"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              step={null}
              type="password"
            />
          </div>
        </div>
      </div>
      <div
        className="css-l5w1vc"
        id="login-child2"
      >
        <button
          className="chakra-button css-3cqj6h"
          id="submit"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
  <div
    aria-atomic={false}
    aria-live="assertive"
    className="css-1oqrfgm"
    data-isinvalid={false}
    id="loginError"
  >
    <div
      className="css-0"
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
  </div>
</div>
`;

exports[`Login Snapshot renders the UI snapshot correctly 2`] = `
<div
  className="css-jgblkz"
>
  <h2
    className="chakra-heading css-1iatin6"
  >
    Log In
  </h2>
  <form
    aria-label="form"
    className="css-1xdhyk6"
    id="login"
    onSubmit={[Function]}
  >
    <div
      className="css-fadkce"
      id="login-parent"
    >
      <div
        className="css-l5w1vc"
        id="login-child0"
      >
        <div
          className="css-5eatfv"
          id="username-wrapper"
        >
          <label
            className="css-1xdhyk6"
            htmlFor="username"
            id="username-label"
          >
            Username
          </label>
          <div
            className="css-79elbk"
          >
            <input
              aria-required={true}
              autoComplete={null}
              className="chakra-input css-0"
              disabled={false}
              id="username"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              step={null}
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        className="css-l5w1vc"
        id="login-child1"
      >
        <div
          className="css-5eatfv"
          id="password-wrapper"
        >
          <label
            className="css-1xdhyk6"
            htmlFor="password"
            id="password-label"
          >
            Password
          </label>
          <div
            className="css-79elbk"
          >
            <input
              aria-required={true}
              autoComplete={null}
              className="chakra-input css-0"
              disabled={false}
              id="password"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              step={null}
              type="password"
            />
          </div>
        </div>
      </div>
      <div
        className="css-l5w1vc"
        id="login-child2"
      >
        <button
          className="chakra-button css-3cqj6h"
          id="submit"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
  <div
    aria-atomic={false}
    aria-live="assertive"
    className="css-1oqrfgm"
    data-isinvalid={false}
    id="loginError"
  >
    <div
      className="css-0"
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
  </div>
</div>
`;
